File: src/contact_pkg/contact_pkg/feasibility_test_node2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import threading
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
from geometry_msgs.msg import WrenchStamped

class FeasibilityTestNode2(Node):
    def __init__(self):
        super().__init__('feasibility_test_node2') 
        
        self.cmd_motion_pub = self.create_publisher(Float64MultiArray, '/ur10skku/cmdMotion', 10)
        
        self.current_p_sub = self.create_subscription(
            Float64MultiArray, '/ur10skku/currentP', self.current_p_callback, 10)
        self.ft_data_sub = self.create_subscription(
            WrenchStamped, '/ur10skku/ftdata', self.ft_data_callback, 10)
            
        self.current_pos = None  
        self.current_ft = None   
        self.home_pos = None     # [추가] 초기 위치를 저장할 변수
        
        self.fixed_wx = 0.0
        self.fixed_wy = 0.0
        self.fixed_wz = 1.57
        self.fixed_fx = 0.0
        self.fixed_fy = 0.0
        self.fixed_fz = 0.0

        self.timer_period = 0.01  
        self.z_step_mm = 0.5      
        
        # 0: 대기, 1: 목표 이동, 2: 수직 하강, 3: 접촉 알고리즘, 4: 원위치 복귀
        self.state = 0 
        
        self.target_x = 0.0
        self.target_y = 0.0
        self.target_wx = 0.0
        self.target_wy = 0.0
        self.target_wz = 1.57
        self.start_z = 0.0 
        # [핵심 추가] 닿는 순간의 위치와 자세를 영구 고정할 변수
        self.locked_pos = None
        # 무한 루프로 입력을 받는 스레드
        self.input_thread = threading.Thread(target=self.wait_for_user_input)
        self.input_thread.daemon = True
        self.input_thread.start()

        self.timer = self.create_timer(self.timer_period, self.control_loop)
        
        self.get_logger().info("피지빌리티 테스트 노드 V2 (원위치 복귀 기능 탑재) 실행 완료.")

    def current_p_callback(self, msg):
        self.current_pos = msg.data

    def ft_data_callback(self, msg):
        fx = msg.wrench.force.x
        fy = msg.wrench.force.y
        fz = msg.wrench.force.z
        mx = msg.wrench.torque.x
        my = msg.wrench.torque.y
        mz = msg.wrench.torque.z
        self.current_ft = [fx, fy, fz, mx, my, mz]

    def wait_for_user_input(self):
        import time
        while self.current_pos is None:
            time.sleep(0.1)

        # [핵심] 최초 1회만 현재 위치를 Home으로 굳혀서 저장합니다.
        self.home_pos = self.current_pos[:]

        # 노드가 살아있는 동안 계속 메뉴를 띄웁니다.
        while rclpy.ok():
            print("\n" + "="*50)
            print(" [로봇 제어 커맨드 라인]")
            print("  s : 새로운 테스트 시작 (목표 위치 지정 후 강하)")
            print("  h : 원위치(Home)로 강제 복귀")
            print("="*50)
            
            cmd = input("명령어 입력 (s 또는 h): ").strip().lower()

            if cmd == 's':
                try:
                    in_x = input("이동할 목표 X 위치 (mm, 그대로 두려면 엔터): ")
                    in_y = input("이동할 목표 Y 위치 (mm, 그대로 두려면 엔터): ")
                    print("\n자세(Orientation) 설정도 변경하시겠습니까? (rad 단위)")
                    in_wx = input(f"목표 WX (기본값 {self.fixed_wx}): ")
                    in_wy = input(f"목표 WY (기본값 {self.fixed_wy}): ")
                    in_wz = input(f"목표 WZ (기본값 {self.fixed_wz}): ")

                    self.target_x = float(in_x) if in_x.strip() != "" else self.current_pos[0]
                    self.target_y = float(in_y) if in_y.strip() != "" else self.current_pos[1]
                    
                    self.target_wx = float(in_wx) if in_wx.strip() != "" else self.fixed_wx
                    self.target_wy = float(in_wy) if in_wy.strip() != "" else self.fixed_wy
                    self.target_wz = float(in_wz) if in_wz.strip() != "" else self.fixed_wz
                    
                    self.start_z = self.current_pos[2] 
                    
                    self.fixed_wx = self.target_wx
                    self.fixed_wy = self.target_wy
                    self.fixed_wz = self.target_wz
                    self.locked_pos = None

                    print(f"\n[명령 접수] X:{self.target_x}, Y:{self.target_y} 로 이동 후 하강합니다.")
                    self.state = 1 # 테스트 시작

                except ValueError:
                    print("입력 오류: 숫자를 정확히 입력해주세요.")
            
            elif cmd == 'h':
                print("\n[인터럽트] 로봇을 최초 초기 위치(Home)로 강제 복귀시킵니다.")
                self.state = 4 # 원위치 복귀 상태로 즉시 전환
            
            else:
                print("알 수 없는 명령어입니다. 's' 또는 'h'를 입력해주세요.")

    def contact_point_detection_algorithm(self, pos, ft):
        self.get_logger().info("접촉 알고리즘 구동 중... (중지하려면 터미널에 h 입력)", throttle_duration_sec=2.0)
        x, y, z = pos[0], pos[1], pos[2]
        return x, y, z 

    def control_loop(self):
        if self.current_pos is None or self.current_ft is None:
            return

        x, y, z, wx, wy, wz = self.current_pos
        _, _, fz, _, _, _ = self.current_ft

        cmd_msg = Float64MultiArray()
        contact_threshold = 2.0 

        if self.state == 0:
            return

        elif self.state == 1:
            error_x = abs(self.target_x - x)
            error_y = abs(self.target_y - y)
            error_wx = abs(self.target_wx - wx)
            error_wy = abs(self.target_wy - wy)
            error_wz = abs(self.target_wz - wz)
            
            if error_x < 1.0 and error_y < 1.0 and error_wx < 0.05 and error_wy < 0.05 and error_wz < 0.05:
                self.get_logger().info("목표 위치 도달 완료! Z축 하강을 시작합니다.")
                self.state = 2 
            else:
                cmd_msg.data = [
                    self.target_x, self.target_y, self.start_z, 
                    self.target_wx, self.target_wy, self.target_wz, 
                    self.fixed_fx, self.fixed_fy, self.fixed_fz
                ]
                self.cmd_motion_pub.publish(cmd_msg)

        elif self.state == 2:
            if abs(fz) > contact_threshold:
                self.get_logger().info(f"표면 접촉 감지! (Fz: {fz:.2f}N). 로봇 위치를 고정합니다.")
                
                # [핵심 변경] 닿는 순간의 6D 포즈(위치+자세) 전체 배열을 '박제'합니다.
                self.locked_pos = self.current_pos[:]
                self.state = 3 
            else:
                target_z = z - self.z_step_mm
                cmd_msg.data = [
                    self.target_x, self.target_y, target_z, 
                    self.fixed_wx, self.fixed_wy, self.fixed_wz, 
                    self.fixed_fx, self.fixed_fy, self.fixed_fz
                ]
                self.cmd_motion_pub.publish(cmd_msg)

        elif self.state == 3:
            # [핵심 변경] 로봇이 떠오르거나 미끄러지지 못하도록, 
            # 어드미턴스 제어기가 밀어내든 말든 무조건 닿았던 순간의 위치(locked_pos)만을 목표치로 계속 쏩니다.
            cmd_msg.data = [
                self.locked_pos[0], self.locked_pos[1], self.locked_pos[2], 
                self.locked_pos[3], self.locked_pos[4], self.locked_pos[5], 
                self.fixed_fx, self.fixed_fy, self.fixed_fz
            ]
            self.cmd_motion_pub.publish(cmd_msg)
            
            # (옵션) 2초마다 정지 중임을 알림
            self.get_logger().info("로봇 정지 및 접촉 유지 중... (원위치 복귀는 터미널에 h 입력)", throttle_duration_sec=2.0)

        # [추가] 상태 4: 저장해둔 Home 위치로 복귀
        elif self.state == 4:
            home_x, home_y, home_z, home_wx, home_wy, home_wz = self.home_pos
            
            error_x = abs(home_x - x)
            error_y = abs(home_y - y)
            error_z = abs(home_z - z)
            
            # 초기 위치 오차가 1mm 미만이면 도착으로 간주
            if error_x < 1.0 and error_y < 1.0 and error_z < 1.0:
                if getattr(self, '_home_reached', False) == False:
                    self.get_logger().info("원위치 복귀 완료! 터미널에서 새 명령(s)을 입력하세요.")
                    self._home_reached = True
                self.state = 0 # 도착 후 다시 대기 상태로 전환
            else:
                self._home_reached = False
                # Home 좌표로 명령 지속 발행
                cmd_msg.data = [
                    home_x, home_y, home_z, 
                    home_wx, home_wy, home_wz, 
                    0.0, 0.0, 0.0
                ]
                self.cmd_motion_pub.publish(cmd_msg)

def main(args=None):
    rclpy.init(args=args)
    node = FeasibilityTestNode2()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("노드가 종료되었습니다.")
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/contact_pkg/src/feasibility_test_node2.cpp
#include <array>
#include <atomic>
#include <chrono>
#include <cmath>
#include <iostream>
#include <memory>
#include <mutex>
#include <optional>
#include <string>
#include <thread>

#include "geometry_msgs/msg/wrench_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"

class FeasibilityTestNode2 : public rclcpp::Node
{
public:
  FeasibilityTestNode2()
  : Node("feasibility_test_node2")
  {
    cmd_motion_pub_ = create_publisher<std_msgs::msg::Float64MultiArray>("/ur10skku/cmdMotion", 10);

    current_p_sub_ = create_subscription<std_msgs::msg::Float64MultiArray>(
      "/ur10skku/currentP", 10,
      [this](const std_msgs::msg::Float64MultiArray::SharedPtr msg) {
        if (msg->data.size() < 6) {
          RCLCPP_WARN_THROTTLE(
            get_logger(), *get_clock(), 2000,
            "currentP 데이터 길이가 부족합니다. expected >= 6, got %zu", msg->data.size());
          return;
        }

        std::lock_guard<std::mutex> lock(data_mutex_);
        for (size_t i = 0; i < current_pos_.size(); ++i) {
          current_pos_[i] = msg->data[i];
        }
        has_current_pos_ = true;
      });

    ft_data_sub_ = create_subscription<geometry_msgs::msg::WrenchStamped>(
      "/ur10skku/ftdata", 10,
      [this](const geometry_msgs::msg::WrenchStamped::SharedPtr msg) {
        std::lock_guard<std::mutex> lock(data_mutex_);
        current_ft_ = {
          msg->wrench.force.x,
          msg->wrench.force.y,
          msg->wrench.force.z,
          msg->wrench.torque.x,
          msg->wrench.torque.y,
          msg->wrench.torque.z};
        has_current_ft_ = true;
      });

    input_thread_ = std::thread(&FeasibilityTestNode2::waitForUserInput, this);

    timer_ = create_wall_timer(
      std::chrono::milliseconds(10),
      std::bind(&FeasibilityTestNode2::controlLoop, this));

    RCLCPP_INFO(get_logger(), "피지빌리티 테스트 노드 V2 (원위치 복귀 기능 탑재) 실행 완료.");
  }

  ~FeasibilityTestNode2() override
  {
    stop_input_.store(true);
    if (input_thread_.joinable()) {
      input_thread_.detach();
    }
  }

private:
  static double parseOrDefault(const std::string & input, double default_value)
  {
    if (input.empty()) {
      return default_value;
    }
    return std::stod(input);
  }

  void waitForUserInput()
  {
    while (rclcpp::ok() && !stop_input_.load()) {
      {
        std::lock_guard<std::mutex> lock(data_mutex_);
        if (has_current_pos_) {
          home_pos_ = current_pos_;
          break;
        }
      }
      std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    while (rclcpp::ok() && !stop_input_.load()) {
      std::cout << "\n" << std::string(50, '=') << "\n";
      std::cout << " [로봇 제어 커맨드 라인]\n";
      std::cout << "  s : 새로운 테스트 시작 (목표 위치 지정 후 강하)\n";
      std::cout << "  h : 원위치(Home)로 강제 복귀\n";
      std::cout << std::string(50, '=') << "\n";

      std::string cmd;
      std::cout << "명령어 입력 (s 또는 h): ";
      if (!std::getline(std::cin, cmd)) {
        return;
      }

      if (cmd == "s") {
        try {
          std::array<double, 6> pos{};
          double current_fixed_wx = 0.0;
          double current_fixed_wy = 0.0;
          double current_fixed_wz = 1.57;
          {
            std::lock_guard<std::mutex> lock(data_mutex_);
            pos = current_pos_;
            current_fixed_wx = fixed_wx_;
            current_fixed_wy = fixed_wy_;
            current_fixed_wz = fixed_wz_;
          }

          std::string in_x;
          std::string in_y;
          std::string in_wx;
          std::string in_wy;
          std::string in_wz;

          std::cout << "이동할 목표 X 위치 (mm, 그대로 두려면 엔터): ";
          std::getline(std::cin, in_x);
          std::cout << "이동할 목표 Y 위치 (mm, 그대로 두려면 엔터): ";
          std::getline(std::cin, in_y);
          std::cout << "\n자세(Orientation) 설정도 변경하시겠습니까? (rad 단위)\n";
          std::cout << "목표 WX (기본값 " << current_fixed_wx << "): ";
          std::getline(std::cin, in_wx);
          std::cout << "목표 WY (기본값 " << current_fixed_wy << "): ";
          std::getline(std::cin, in_wy);
          std::cout << "목표 WZ (기본값 " << current_fixed_wz << "): ";
          std::getline(std::cin, in_wz);

          const double new_target_x = parseOrDefault(in_x, pos[0]);
          const double new_target_y = parseOrDefault(in_y, pos[1]);
          const double new_target_wx = parseOrDefault(in_wx, current_fixed_wx);
          const double new_target_wy = parseOrDefault(in_wy, current_fixed_wy);
          const double new_target_wz = parseOrDefault(in_wz, current_fixed_wz);

          {
            std::lock_guard<std::mutex> lock(data_mutex_);
            target_x_ = new_target_x;
            target_y_ = new_target_y;
            target_wx_ = new_target_wx;
            target_wy_ = new_target_wy;
            target_wz_ = new_target_wz;
            start_z_ = current_pos_[2];
            fixed_wx_ = target_wx_;
            fixed_wy_ = target_wy_;
            fixed_wz_ = target_wz_;
            locked_pos_.reset();
            state_ = 1;
          }

          std::cout << "\n[명령 접수] X:" << new_target_x
                    << ", Y:" << new_target_y << " 로 이동 후 하강합니다.\n";
        } catch (const std::exception &) {
          std::cout << "입력 오류: 숫자를 정확히 입력해주세요.\n";
        }
      } else if (cmd == "h") {
        std::cout << "\n[인터럽트] 로봇을 최초 초기 위치(Home)로 강제 복귀시킵니다.\n";
        std::lock_guard<std::mutex> lock(data_mutex_);
        state_ = 4;
      } else {
        std::cout << "알 수 없는 명령어입니다. 's' 또는 'h'를 입력해주세요.\n";
      }
    }
  }

  void controlLoop()
  {
    std::array<double, 6> current_pos{};
    std::array<double, 6> current_ft{};
    std::array<double, 6> home_pos{};
    std::optional<std::array<double, 6>> locked_pos;
    int state = 0;
    double target_x = 0.0;
    double target_y = 0.0;
    double target_wx = 0.0;
    double target_wy = 0.0;
    double target_wz = 1.57;
    double start_z = 0.0;
    double fixed_wx = 0.0;
    double fixed_wy = 0.0;
    double fixed_wz = 1.57;
    double fixed_fx = 0.0;
    double fixed_fy = 0.0;
    double fixed_fz = 10.0;
    bool has_home_pos = false;

    {
      std::lock_guard<std::mutex> lock(data_mutex_);
      if (!has_current_pos_ || !has_current_ft_) {
        return;
      }

      current_pos = current_pos_;
      current_ft = current_ft_;
      if (home_pos_.has_value()) {
        home_pos = home_pos_.value();
        has_home_pos = true;
      }
      locked_pos = locked_pos_;
      state = state_;
      target_x = target_x_;
      target_y = target_y_;
      target_wx = target_wx_;
      target_wy = target_wy_;
      target_wz = target_wz_;
      start_z = start_z_;
      fixed_wx = fixed_wx_;
      fixed_wy = fixed_wy_;
      fixed_wz = fixed_wz_;
      fixed_fx = fixed_fx_;
      fixed_fy = fixed_fy_;
      fixed_fz = fixed_fz_;
    }

    const double x = current_pos[0];
    const double y = current_pos[1];
    const double z = current_pos[2];
    const double wx = current_pos[3];
    const double wy = current_pos[4];
    const double wz = current_pos[5];
    const double fz = current_ft[2];

    std_msgs::msg::Float64MultiArray cmd_msg;

    if (state == 0) {
      return;
    }

    if (state == 1) {
      const double error_x = std::abs(target_x - x);
      const double error_y = std::abs(target_y - y);
      const double error_wx = std::abs(target_wx - wx);
      const double error_wy = std::abs(target_wy - wy);
      const double error_wz = std::abs(target_wz - wz);

      if (error_x < 1.0 && error_y < 1.0 &&
        error_wx < 0.05 && error_wy < 0.05 && error_wz < 0.05)
      {
        RCLCPP_INFO(get_logger(), "목표 위치 도달 완료! Z축 하강을 시작합니다.");
        std::lock_guard<std::mutex> lock(data_mutex_);
        state_ = 2;
      } else {
        cmd_msg.data = {
          target_x, target_y, start_z,
          target_wx, target_wy, target_wz,
          fixed_fx, fixed_fy, fixed_fz};
        cmd_motion_pub_->publish(cmd_msg);
      }
      return;
    }

    if (state == 2) {
      if (std::abs(fz) > contact_threshold_) {
        RCLCPP_INFO(get_logger(), "표면 접촉 감지! (Fz: %.2fN). 로봇 위치를 고정합니다.", fz);
        std::lock_guard<std::mutex> lock(data_mutex_);
        locked_pos_ = current_pos_;
        state_ = 3;
      } else {
        const double target_z = z - z_step_mm_;
        cmd_msg.data = {
          target_x, target_y, target_z,
          fixed_wx, fixed_wy, fixed_wz,
          fixed_fx, fixed_fy, fixed_fz};
        cmd_motion_pub_->publish(cmd_msg);
      }
      return;
    }

    if (state == 3) {
      if (!locked_pos.has_value()) {
        return;
      }

      cmd_msg.data = {
        locked_pos->at(0), locked_pos->at(1), locked_pos->at(2),
        locked_pos->at(3), locked_pos->at(4), locked_pos->at(5),
        fixed_fx, fixed_fy, fixed_fz};
      cmd_motion_pub_->publish(cmd_msg);
      RCLCPP_INFO_THROTTLE(
        get_logger(), *get_clock(), 2000,
        "로봇 정지 및 접촉 유지 중... (원위치 복귀는 터미널에 h 입력)");
      return;
    }

    if (state == 4) {
      if (!has_home_pos) {
        return;
      }

      const double error_x = std::abs(home_pos[0] - x);
      const double error_y = std::abs(home_pos[1] - y);
      const double error_z = std::abs(home_pos[2] - z);

      if (error_x < 1.0 && error_y < 1.0 && error_z < 1.0) {
        std::lock_guard<std::mutex> lock(data_mutex_);
        if (!home_reached_) {
          RCLCPP_INFO(get_logger(), "원위치 복귀 완료! 터미널에서 새 명령(s)을 입력하세요.");
          home_reached_ = true;
        }
        state_ = 0;
      } else {
        {
          std::lock_guard<std::mutex> lock(data_mutex_);
          home_reached_ = false;
        }
        cmd_msg.data = {
          home_pos[0], home_pos[1], home_pos[2],
          home_pos[3], home_pos[4], home_pos[5],
          0.0, 0.0, 0.0};
        cmd_motion_pub_->publish(cmd_msg);
      }
    }
  }

  rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr cmd_motion_pub_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr current_p_sub_;
  rclcpp::Subscription<geometry_msgs::msg::WrenchStamped>::SharedPtr ft_data_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::thread input_thread_;
  std::atomic_bool stop_input_{false};
  std::mutex data_mutex_;

  std::array<double, 6> current_pos_{};
  std::array<double, 6> current_ft_{};
  std::optional<std::array<double, 6>> home_pos_;
  std::optional<std::array<double, 6>> locked_pos_;
  bool has_current_pos_{false};
  bool has_current_ft_{false};
  bool home_reached_{false};

  int state_{0};
  double target_x_{0.0};
  double target_y_{0.0};
  double target_wx_{0.0};
  double target_wy_{0.0};
  double target_wz_{1.57};
  double start_z_{0.0};
  double fixed_wx_{0.0};
  double fixed_wy_{0.0};
  double fixed_wz_{1.57};
  double fixed_fx_{0.0};
  double fixed_fy_{0.0};
  double fixed_fz_{10.0};

  const double z_step_mm_{0.5};
  const double contact_threshold_{2.0};
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<FeasibilityTestNode2>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
